lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+package lib
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"github.com/deso-protocol/go-deadlock"
+	"github.com/golang/glog"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"os"
+	"sync"
+	"time"
+)
+
+// StateSyncerOperationType is an enum that represents the type of operation that should be performed on the
+// state consumer database.
+type StateSyncerOperationType uint8
+
+const (
+	DbOperationTypeInsert StateSyncerOperationType = 0
+	DbOperationTypeDelete StateSyncerOperationType = 1
+	DbOperationTypeUpsert StateSyncerOperationType = 2
+)
+
+// StateChangeEntry is used to capture the state of the database. These changes are then written to a file, which is
+// then used to sync data consumers who subscribe to changes to that file.
+type StateChangeEntry struct {
+	// The type of operation that should be performed on the database.
+	OperationType StateSyncerOperationType
+	// The key that should be used for the operation.
+	KeyBytes []byte
+	// The encoder that is being captured by this state change.
+	Encoder DeSoEncoder
+	// The encoder represented in bytes. This could be created by just calling EncodeToBytes on the encoder, but
+	// during operations like hypersync, we are given the raw bytes of the encoder, which is all we need to encode the
+	// StateChangeEntry. Thus, we store the raw bytes here to avoid having to re-encode the encoder.
+	EncoderBytes []byte
+	// The ancestral record that should be used to revert this transaction.
+	AncestralRecord DeSoEncoder
+	// The ancestral record represented in bytes.
+	AncestralRecordBytes []byte
```

Comment:
Add comment on when ancestral record bytes is used. I assume in hypersync it's not pupulated for example. Also I think it's only for upsert?

Reply:
Yep, it's only for upsert. We only use ancestral records when syncing mempool entries. When the new block syncs, we need to revert any mempool entries we've stored in the database before applying the committed entries. We do this reversion by reverting to the ancestral record value.